### ISSUE TITLE
Upgrade mysql jdbc jar

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#1589]
The mysql jdbc jar should be upgraded.  This was attempted before, causing multiple issues:
MIRTH-1034
MIRTH-1046
MIRTH-1047

DatabaseConnection.java was just changed, which should resolve the issues with MIRTH-1034 and MIRTH-1046.  The issues with MIRTH-1047 still need to be tested and resolved for an upgrade to be possible.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-1594
Reporter: jacobb
Created: 2010-10-14T11:29:03.000-0700

**Comment by rbeckman-nextgen** [comment]
Upgrading from a JDBC 3 to a JDBC 4 connector would simplify some setBinaryStream calls.  I believe it would eliminate the need to pass the length of the blob as a parameter.  That would be valuable to me.

Imported Comment. Original Details:
Author: shawn
Created: 2012-02-08T20:07:17.000-0800

**Comment by rbeckman-nextgen** [comment]
Regarding MIRTH-1046:

I tested inserting and returning generated keys on a table using an AI column with four methods.

Method 1: Create the statement and call getGeneratedKeys() without requesting them beforehand

Method 2: Create the statement passing in RETURN_GENERATED_KEYS, and then calling getGeneratedKeys()

Method 3: Create the statement passing in an array of column names to return keys for, and then calling getGeneratedKeys()

Method 4: Use DatabaseConnection.executeUpdateAndGetGeneratedKeys()

I tested this with MySQL 5.1.30 and 5.5.28, and Connector/J 5.0.8, 5.1.7, and 5.1.25 (the latest). Here's what I found:

||Connector/J Version||MySQL Version||Method 1||Method 2||Method 3||Method 4||
|5.0.8|5.1.30|Y|Y|Y|Y|
|5.1.7|5.1.30|N|Y|Y|Y|
|5.1.25|5.1.30|N|Y|Y|Y|
|5.0.8|5.5.28|Y|Y|Y|Y|
|5.1.7|5.5.28|N|Y|Y|Y|
|5.1.25|5.5.28|N|Y|Y|Y|

"Y" indicates that the method worked, and "N" indicates that an exception happened. When you call getGeneratedKeys(), you're _**supposed**_ to have requested that beforehand, either by passing in RETURN_GENERATED_KEYS or by passing in an array of column indices or names. If you don't, then you will get the "GeneratedKeysNotRequested" exception. That is _expected_. So Method 1 in the above table should be "N" across the board, but it's actually "Y" for 5.0.8. MIRTH-1046 seems to indicate that it was a new bug that occurred in 5.1.7 and later, but it's actually the other way around. The bug was in 5.0.8, and it was fixed in later versions. Here's the related issue: [http://bugs.mysql.com/bug.php?id=34185]

In the [issue posted on MySQL's tracker](http://bugs.mysql.com/bug.php?id=41448), one of the users said that he saw the exception happen even when he passed in RETURN_GENERATED_KEYS. I was **not** able to reproduce this (Method 2, 3 and 4 in the table above are "Y" across the board). I even used the specific versions of MySQL and Connector/J that he used and still wasn't able to reproduce it. Reading through the issue comments, I wouldn't be surprised if he did something else wrong in his code to cause the exception, or he copied/pasted wrong, or lost track of which cause went with which effect, or something like that.

Bottom line: If any users are currently using JavaScript in Mirth Connect to call getGeneratedKeys() on a MySQL statement without having requested them first, that **will** break if we upgrade to any later version. But that's actually correct; it _should_ break, because they're doing it wrong.

Imported Comment. Original Details:
Author: narupley
Created: 2013-07-16T09:02:09.000-0700

**Comment by rbeckman-nextgen** [comment]
Regarding MIRTH-1047:

I tested with Connector/J 5.0.8 and 5.1.6 and verified that both work as expected. I suspect this may be another case of the user just not following the API correctly. From [Statement](http://docs.oracle.com/javase/1.4.2/docs/api/java/sql/Statement.html):

> By default, only one ResultSet object per Statement object can be open at the same time. Therefore, if the reading of one ResultSet object is interleaved with the reading of another, each must have been generated by different Statement objects. All execution methods in the Statement interface implicitly close a statment's current ResultSet object if an open one exists.


Multiple result sets _are_ allowed, just not on the same Statement. If you were attempting to use multiple result sets from _multiple_ statements and you _still_ got the "Operation not allowed after ResultSet closed" exception, that would be a different story. But I have not seen that on any permutation of Connector/J / MySQL versions that I've tested with. DatabaseConnection.executeCachedQuery() should not even play a role in this issue because it's always using it's own self-created Statement.

If MIRTH-1047 arose because of using multiple result sets spawned from multiple statements (and I doubt it did), then I would need a complete test case supplied, because I'm not able to reproduce the error in any version while following the API correctly.

Imported Comment. Original Details:
Author: narupley
Created: 2013-07-16T10:05:26.000-0700

**Comment by rbeckman-nextgen** [comment]
Revision 6670: Updated MySQL Connector/J to version 5.1.25.

Imported Comment. Original Details:
Author: narupley
Created: 2013-07-16T15:05:28.000-0700

**Comment by rbeckman-nextgen** [comment]
Verified Nick's comments and checked that executeCachedQuery() works.

Imported Comment. Original Details:
Author: wayneh
Created: 2013-08-12T11:15:59.000-0700
